Libraries/LLM/Llama.swift

```diff
@@ -17,43 +97,40 @@ private class Attention: Module {
     @ModuleInfo(key: "v_proj") var wv: Linear
     @ModuleInfo(key: "o_proj") var wo: Linear
 
-    let rope: RoPE
+    let rope: DynamicNTKScalingRoPE
 
-    public init(_ args: LlamaConfiguration) {
+    init(_ args: LlamaConfiguration) {
         self.args = args
 
         let dim = args.hiddenSize
         let heads = args.attentionHeads
         let kvHeads = args.kvHeads
 
-        let headDim = args.hiddenSize / heads
+        let headDim = args.headDimensions ?? (args.hiddenSize / heads)
         self.scale = pow(Float(headDim), -0.5)
 
-        self._wq.wrappedValue = Linear(dim, heads * headDim, bias: false)
-        self._wk.wrappedValue = Linear(dim, kvHeads * headDim, bias: false)
-        self._wv.wrappedValue = Linear(dim, kvHeads * headDim, bias: false)
-        self._wo.wrappedValue = Linear(heads * headDim, dim, bias: false)
-
-        let ropeScale: Float
-        if let ropeScaling = args.ropeScaling, ropeScaling["type"] == .string("linear"),
-            let factor = ropeScaling["factor"]
-        {
-            switch factor {
-            case .string:
-                fatalError("ropeScaling.factor must be a float")
-            case .float(let v):
-                ropeScale = 1 / v
-            }
-        } else {
-            ropeScale = 1
-        }
-
-        self.rope = RoPE(
-            dimensions: headDim, traditional: args.ropeTraditional, base: args.ropeTheta,
-            scale: ropeScale)
+        self._wq.wrappedValue = Linear(dim, heads * headDim, bias: args.attentionBias)
+        self._wk.wrappedValue = Linear(dim, kvHeads * headDim, bias: args.attentionBias)
+        self._wv.wrappedValue = Linear(dim, kvHeads * headDim, bias: args.attentionBias)
+        self._wo.wrappedValue = Linear(heads * headDim, dim, bias: args.attentionBias)
+
+        self.rope = DynamicNTKScalingRoPE(
+            dims: headDim,
+            maxPositionEmbeddings: args.maxPositionEmbeddings ?? 2048,
```

Comment:
Maybe it's better to have `maxPositionEmbeddings` be an optional value in `DynamicNTKScalingRoPE` instead of arbitrarily setting it to 2048 (which I think is wrong). Also because you test it for `> 0` so we will scale the `base` for models that we don't want to scale for.